lib/utils/removeUselessVias.ts

```diff
@@ -0,0 +1,113 @@
+import { HighDensityIntraNodeRoute } from "lib/types/high-density-types"
+import { distance, doSegmentsIntersect } from "@tscircuit/math-utils"
+
+interface Point3D {
+  x: number
+  y: number
+  z: number
+}
+
+/**
+ * Checks if a path between two points on the same layer intersects with any obstacles
+ */
+function pathIntersectsObstacles(
+  start: Point3D,
+  end: Point3D,
+  obstacleRoutes: HighDensityIntraNodeRoute[],
+  traceThickness: number,
+  viaDiameter: number,
+): boolean {
+  if (start.z !== end.z) return false
+
+  for (const route of obstacleRoutes) {
+    // Check intersection with traces
+    for (let i = 0; i < route.route.length - 1; i++) {
+      const routeStart = route.route[i]
+      const routeEnd = route.route[i + 1]
+
+      // Only check segments on the same layer
+      if (routeStart.z === routeEnd.z && routeStart.z === start.z) {
+        if (doSegmentsIntersect(start, end, routeStart, routeEnd)) {
+          return true
+        }
+      }
+    }
+
+    // Check intersection with vias
+    for (const via of route.vias) {
+      const viaPoint = { ...via, z: start.z }
+      const distToVia = Math.min(
+        distance(start, viaPoint),
+        distance(end, viaPoint),
+      )
+      if (distToVia < (viaDiameter + traceThickness) / 2) {
+        return true
+      }
+    }
+  }
+
+  return false
+}
+
+/**
+ * Removes unnecessary vias from a route. A via is considered unnecessary if:
+ * 1. It changes layers but immediately changes back to the original layer
+ * 2. It changes layers when the entire path could stay on a single layer without intersecting obstacles
```

Comment:
Maybe more a question for @seveibar, but I wonder if it's possible that multiple traces have a via to avoid each other while the number of vias could be reduced by removing multiple vias in one go.